scanner/decisions/0010-separate-versioning-for-vuln-bundles.md

```diff
@@ -0,0 +1,175 @@
+# 0010 - Separate Versioning Stream for Vulnerability Bundles
+
+- **Author(s):** J. Victor Martins <jvdm@sdf.org>
+- **Created:** [2024-07-22 Mon]
+
+## Status
+
+Status: Accepted
+
+Updates: [#0009](0009-scannerv4-read-manual-vulnerabilities.md)
+
+## Context
+
+A Vulnerability Bundle is an archive containing updates to vulnerability data in Scanner V4 (aka. Scanner).  They are constantly updated by the Vulnerability Exporter (aka. Exporter), which runs periodically in GitHub Workflows and uses ClairCore to fetch the security data to convert it to a JSON-based schema.  Then, the workflow bundles the data into a ZIP archive.  ClairCore owns the schema.
+
+The Scanner Matcher, using ClairCore, runs the Vulnerability Updater (aka. Updater), which fetches the bundle, parses the data, and updates the vulnerability database.
+
+Suppose a backward-incompatible schema change is made in ClairCore and incorporated into Scanner.  In that case, the following Exporter execution will update the bundle adhering to the new schema.  Existing Scanner instances can only parse the changed data if they upgrade to the new ClairCore version that understands the schema changes.
+
+Due to the tight coupling between the ClairCore version and schema, one bundle per StackRox release was introduced for Scanner V4.
+
+Each bundle is created with an Exporter running the same ClairCore version used to build the release, ensuring compatibility.  The Exporter serves bundles at `https://definitions.stackrox.io/v4/<version>/vulnerabilities.zip` where `<version>` is the supported StackRox release (such as "4.5.0").  Scanner uses its release version tagged during build time to determine which URL to fetch from.
+
+In offline mode, one offline bundle per release was introduced for Scanner V4, coexisting with the "latest" Scanner V2 offline bundles.  To facilitate user migration to per-release offline bundles, the latest offline bundle used by Scanner V2 (at the time of this writing served at https://install.stackrox.io/scanner/scanner-vuln-updates.zip) was changed to contain Scanner V4's offline bundles for `4.[45][.x]` releases.  A command `roxctl scanner download-db` was created to facilitate users fetching the right offline bundle for their release.
+
+This approach needs to be revised for the following reasons:
+
+1.  The vulnerability schema rarely changes, and maintaining multiple bundles is expensive.  It requires multiple GitHub workflow runs, which increases the chances of failure and pressure on the security sources' endpoints.
```

Comment:
Thinking about it some more another factor would be the removal of an updater. For example, the `rhel-vex` updater will replace all `RHELX-XXX` updaters. These vulnerabilities may not be properly matched with older versions of Scanner V4 (the matching logic changed to accommodate the update), so I don't think we can use the VEX data with older Scanner V4 versions. So, this may be an example where the schema did not change, but there was a significant enough change which will cause us to bump versions.

Reply:
That's valid. Unless we keep both data in the bundle, and the `updater.Import()` to ignore OVAL RHEL updaters. Feels like bumping the version a cleaner way to do this.